memory_view.c

```diff
@@ -0,0 +1,447 @@
+{
+    ssize_t i, n = item_size;
+    if (row_major_p) {
+        for (i = ndim - 1; i >= 0; --i) {
+            strides[i] = n;
+            n *= shape[i];
+        }
+    }
+    else {  // column-major
+        for (i = 0; i < ndim; ++i) {
+            strides[i] = n;
+            n *= shape[i];
+        }
+    }
+}
+
+/* Initialize view to expose a simple byte array */
+int
+rb_memory_view_init_as_byte_array(rb_memory_view_t *view, VALUE obj, void *data, const ssize_t len, const int readonly)
+{
+    view->obj = obj;
+    view->data = data;
+    view->len = len;
+    view->readonly = readonly;
+    view->format = NULL;
+    view->item_size = 1;
+    view->ndim = 1;
+    view->shape = NULL;
+    view->strides = NULL;
+    view->sub_offsets = NULL;
+    *((void **)&view->private) = NULL;
+
+    return 1;
+}
+
+static ssize_t
+get_format_size(const char *format, bool *native_p, ssize_t *alignment)
+{
+    RUBY_ASSERT(format != NULL);
+    RUBY_ASSERT(native_p != NULL);
+
+    *native_p = false;
+
+    switch (*format) {
+      case 'x':  // padding
+        return 1;
+
+      case 'c':  // signed char
+      case 'C':  // unsigned char
+        return sizeof(char);
+
+      case 's':  // s for int16_t, s! for signed short
+      case 'S':  // S for uint16_t, S! for unsigned short
+        if (format[1] == '!') {
+            *native_p = true;
+            *alignment = SHORT_ALIGNMENT;
+            return sizeof(short);
+        }
+        // fall through
+
+      case 'n':  // n for big-endian 16bit unsigned integer
+      case 'v':  // v for little-endian 16bit unsigned integer
+        *alignment = INT16_ALIGNMENT;
+        return 2;
+
+      case 'i':  // i and i! for signed int
+      case 'I':  // I and I! for unsigned int
+        *native_p = (format[1] == '!');
+        *alignment = INT_ALIGNMENT;
+        return sizeof(int);
+
+      case 'l':  // l for int32_t, l! for signed long
+      case 'L':  // L for uint32_t, L! for unsigned long
+        if (format[1] == '!') {
+            *native_p = true;
+            *alignment = LONG_ALIGNMENT;
+            return sizeof(long);
+        }
+        // fall through
+
+      case 'N':  // N for big-endian 32bit unsigned integer
+      case 'V':  // V for little-endian 32bit unsigned integer
+      case 'f':  // f for native float
+      case 'e':  // e for little-endian float
+      case 'g':  // g for big-endian float
+        *alignment = INT32_ALIGNMENT;
+        return 4;
+
+      case 'q':  // q for int64_t, q! for signed long long
+      case 'Q':  // Q for uint64_t, Q! for unsigned long long
+        if (format[1] == '!') {
+            *native_p = true;
+            *alignment = LONG_LONG_ALIGNMENT;
+            return sizeof(LONG_LONG);
+        }
+        // fall through
+
+      case 'd':  // d for native double
+      case 'E':  // E for little-endian double
+      case 'G':  // G for big-endian double
+        *alignment = INT64_ALIGNMENT;
```

Comment:
Oops.  Current code should work on i386 because
the alignment of double and 64bit integer are equal.
However, I'm not sure that it is guaranteed for all architectures.

Reply:
I'll use `RUBY_ALIGNOF(double)` for `double`.